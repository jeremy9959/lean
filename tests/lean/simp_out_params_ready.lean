namespace in_param

class A (α β : Type) -- no out_param!
instance inst : A ℕ ℤ := ⟨⟩
def p (_ : Type) : Prop := false
@[simp] lemma l {α β} [A α β] : p α ↔ p β := iff.rfl

-- `simp` should fail to apply `l` here since the `β` parameter to `l` can't be inferred,
-- instead of trying to synthesize an instance `A ℕ ?m_1`, find `inst` and unify `?m_1 =?= ℤ`,
-- turning the goal into `p ℤ`
example : p ℕ := by simp

end in_param

namespace out_param

-- However, if `β` is marked as an out_param, the goal should turn into `p ℤ`
class A (α : Type) (β : out_param Type) -- no out_param!
instance inst : A ℕ ℤ := ⟨⟩
def p (_ : Type) : Prop := false
@[simp] lemma l {α β} [A α β] : p α ↔ p β := iff.rfl
example : p ℕ := by simp

end out_param
