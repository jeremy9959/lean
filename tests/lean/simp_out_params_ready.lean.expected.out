simp_out_params_ready.lean:15:20: error: simplify tactic failed to simplify
state:
⊢ p ℕ
simp_out_params_ready.lean:26:20: error: tactic failed, there are unsolved goals
state:
⊢ p ℤ
